### PR TITLE
Added suitable auto app switching in case of application deletion

### LINKED ISSUE #1331: Active context does not switch after deleting the active application
[kind/bug]

## What versions of software are you using?
- Operating System: Fedora
- Output of `odo version`: master

## How did you run odo exactly?
1. Clone https://github.com/redhat-developer/odo
2. make bin
3. ./odo login -u developer -p password
4. ./odo project create testproject
6. ./odo app create test1
7. ./odo app create test2
8. ./odo app list 
```
$ odo app list
The project 'testproject' has the following applications:
ACTIVE     NAME
           test1
*          test2
```
9. ./odo app delete test2
10. ./odo app list
```
$ odo app list
The project 'testproject' has the following applications:
ACTIVE     NAME
           test1
```
## Actual behavior
Active context does not switch to other available app

## Expected behavior
Active context should switch
```
$ odo app list
The project 'testproject' has the following applications:
ACTIVE     NAME
*          test1
```

### PR DESCRIPTION
**Pull request by mik-dass**
fixes #1331 

How to test:
1. Create many apps
2. Delete the active one and check app switching

### CODE REVIEW

**Comment by amitkrout** [comment]
Ping @geoand @metacosm @girishramnani for review.

@mik-dass Did you check it locally ? i am asking because no integration test is tracking the change. Would be great if you add the integration test for it or else create a follow up issue for integration test.

**Comment by amitkrout** [comment]
@mik-dass Please address the issue #1340 in this pr too. Could be a similar kind of fix i guess.

**Comment by mik-dass** [comment]
> i am asking because no integration test is tracking the change

Ok. But I am confused in which file to add this integration test. WDYT?

**Comment by kadel** [comment]
> @mik-dass Please address the issue #1340 in this pr too. Could be a similar kind of fix i guess.

let's do this in separate PR.

Small PRs that are quick and easy to merge, that is our motto ;-)

File: pkg/config/config.go [hunk 1]
```diff
@@ -555,21 +555,35 @@ func (c *GlobalConfigInfo) DeleteApplication(application string, project string)
...
 	}
 
-	if !found {
+	if foundAt == -1 {
 		return fmt.Errorf("application %s doesn't exist", application)
 
 	}
 
+	// if the deleted app is the active application then set the next app in the project as active
+	if nextFoundAt != -1 && isDeletedAppActive {
```
Comment (mohammedzee1000) [inline pkg/config/config.go]: +1 this should ensure that if application being deleted is the last application, then no active application is set

**Comment by mik-dass** [comment]
> Can we have a test to ensure that if last application is being deleted, then there is no active application set.

Sure I will add that

**Comment by syamgk** [comment]
@mik-dass for adding e2e test https://github.com/redhat-developer/odo/blob/master/tests/e2e/e2e_test.go#L70 this will be the right place 
apart from that it looks good to me

**Comment by mik-dass** [comment]
> @mik-dass for adding e2e test https://github.com/redhat-developer/odo/blob/master/tests/e2e/e2e_test.go#L70 this will be the right place
apart from that it looks good to me

Thanks, I will have a look ASAP :)

File: tests/e2e/utils.go [hunk 2]
```diff
@@ -74,3 +75,18 @@ func cleanUpAfterProjects(projects []string) {
 		odoDeleteProject(p)
 	}
 }
+
+// getActiveApplication returns the active application in the project
+// returns empty string if no active application is present in the project
+func getActiveApplication() string {
```
Comment (amitkrout) [inline tests/e2e/utils.go]: Lovely

File: tests/e2e/e2e_test.go [hunk 3]
```diff
@@ -768,6 +767,38 @@ var _ = Describe("odoe2e", func() {
...
+
+			// delete in the active project
+			// the current app should switch
+			runCmdShouldPass("odo app delete -f")
+			Expect(getActiveApplication()).To(Equal("app-4"))
+
+			// deleting the last app in the active project
+			runCmdShouldPass("odo app delete -f")
+			Expect(getActiveApplication()).To(Equal(""))
+
```
Comment (amitkrout) [inline tests/e2e/e2e_test.go]: You can add few steps to check in project ```-auto-0``` app-3 is still the active app.

For example
```
odo project -auto-0
Expect(getActiveApplication()).To(Equal("app-3"))
```

File: tests/e2e/utils.go (hunk 2 above)
Comment (mik-dass) [reply tests/e2e/utils.go]: What are the changes needed here?

Comment (amitkrout) [reply tests/e2e/utils.go]: haha...
It went to wrong place ;)

File: tests/e2e/e2e_test.go (hunk 3 above)
Comment (mik-dass) [reply tests/e2e/e2e_test.go]: I have created a issue https://github.com/redhat-developer/odo/issues/1363 to track of it

Comment (amitkrout) [reply tests/e2e/e2e_test.go]: Thanks for the issue.

**Pull request merged**